### PR TITLE
network: fetch blocks in parallel

### LINKED ISSUE #1376: Block fetching optimization
At the moment at the height of H we send `GetBlockByIndex(H+1, -1)` requests to all our peers which is suboptimal as each of them sends us the same blocks in response. We should be smarter than that and ask for different blocks, going beyond the default 500 blocks requests (like `GetBlockByIndex(H+1, -1)` to peer A, `GetBlockByIndex(H+500+1, -1)` to peer B, `GetBlockByIndex(H+1000+1, -1)` to peer C, etc). There still should be some redundancy though as the other side can ignore the request. Peer's height should also be taken into account.

But in the end we should fill our block queue more quickly speeding up the synchronization and reducing useless traffic.

### PR DESCRIPTION
**Pull request by fyrchik**
Blockcache size is 2000, while max request size is 500.
Try to fetch blocks in chunks starting from current height.
Lower height has priority.

Close #1376 .

### CODE REVIEW

**Comment by roman-khimov** [comment]
Maybe it can be simplified. You can store the highest requested block in `Server` (`requestedHeight`) and upon entering into `requestBlocks` compare it to `BlockHeight()`. Then
```
if requestedHeight < blockHeight + 1500
    CompareAndSwapUint32(requestedHeight, old, old + 500)
    NewGetBlockByIndex(requestHeight+1, -1)
else
    index := random(4)
    NewGetBlockByIndex(blockHeight+index*500, -1)
```
This way we start with `blockHeight`, then move on to `blockHeight + 500`, then `+ 1000`, then `+ 1500`. If we keep entering this function before `blockHeight` moves `+500`, we're re-requesting _randomly_ things in `blockHeight/blockHeight + 200` range. Non-random requests can lead to starving if we only have 4 peers and one of them never responds for some reason.

And that's it, I think it's a little simpler (no locking, no maps).

**Comment by fyrchik** [comment]
Actually, this solution explicitly prevents starving by not updating number of sent requests if the request sent to client is sent again. Though there are some other unlikely situations when starving is probable (should be easy to fix).

`random` solution is a bit simpler, but harder to test. I'll try it.

File: pkg/network/server.go [hunk 1]
```diff
@@ -710,9 +714,33 @@ func (s *Server) handleGetAddrCmd(p Peer) error {
...
+// 3. After all requests were sent, request random height.
 func (s *Server) requestBlocks(p Peer) error {
-	payload := payload.NewGetBlockByIndex(s.chain.BlockHeight()+1, -1)
+	var currHeight = s.chain.BlockHeight()
+	var peerHeight = p.LastBlockIndex()
+	var needHeight uint32
+	var old = s.lastRequestedHeight.Load()
+	if old <= currHeight {
+		needHeight = currHeight + 1
+		s.lastRequestedHeight.CAS(old, needHeight)
```
Comment (roman-khimov) [inline pkg/network/server.go:732]: CAS returns a value and you need to handle it to ensure correctness.

File: pkg/network/server.go [hunk 2]
```diff
@@ -710,9 +714,33 @@ func (s *Server) handleGetAddrCmd(p Peer) error {
...
+	} else if old < currHeight+1500 {
```
Comment (roman-khimov) [inline pkg/network/server.go:733]: `1500` is magic, use constants for that.

File: pkg/network/server.go [hunk 3]
```diff
@@ -710,9 +714,33 @@ func (s *Server) handleGetAddrCmd(p Peer) error {
...
+		needHeight = currHeight + 1
+		if peerHeight > old+500 {
+			needHeight = old + 500
+			s.lastRequestedHeight.CAS(old, needHeight)
+		}
+	} else {
+		index := mrand.Intn(4)
```
Comment (roman-khimov) [inline pkg/network/server.go:740]: 4 is 2000/500.

**Review by roman-khimov — CHANGES_REQUESTED** [review]

File: pkg/network/server.go (hunk 1 above)
Comment (fyrchik) [reply pkg/network/server.go:732]: Fixed.

File: pkg/network/server.go (hunk 2 above)
Comment (fyrchik) [reply pkg/network/server.go:733]: Fixed.

File: pkg/network/server.go (hunk 3 above)
Comment (fyrchik) [reply pkg/network/server.go:740]: Fixed.

File: pkg/network/server.go [hunk 4]
```diff
@@ -710,9 +714,41 @@ func (s *Server) handleGetAddrCmd(p Peer) error {
...
+			needHeight = currHeight + 1
+			if peerHeight > old+payload.MaxHashesCount {
+				needHeight = old + payload.MaxHashesCount
+				if !s.lastRequestedHeight.CAS(old, needHeight) {
+					continue
+				}
+			}
+		} else {
+			index := mrand.Intn(blockCacheSize / payload.MaxHashesCount)
+			needHeight = currHeight + 1 + uint32(index*500)
```
Comment (roman-khimov) [inline pkg/network/server.go:747]: `MaxHashesCount`

**Review by AnnaShaleva — APPROVED** [review]

**Review by roman-khimov — APPROVED** [review]

**Pull request merged**